framework/src/userobject/ActivateElementUserObject.C

```diff
@@ -0,0 +1,536 @@
+  params.addClassDescription("Determine activated elements.");
+  params.addRequiredParam<subdomain_id_type>("active_subdomain_id", "The active subdomain ID.");
+  params.addParam<subdomain_id_type>(
+      "inactive_subdomain_id",
+      -1,
+      "The inactivate subdomain ID, i.e., the subdomain that you want to keep the same.");
+  params.addRequiredParam<std::vector<BoundaryName>>("expand_boundary_name",
+                                                     "The expanded boundary name.");
+  params.addParam<FunctionName>("function_x", "The x component heating spot travel path");
+  params.addParam<FunctionName>("function_y", "The y component heating spot travel path");
+  params.addParam<FunctionName>("function_z", "The z component heating spot travel path");
+  params.addParam<Real>("activate_distance",
+                        1e-4,
+                        "The maximum distance of the activated element to the point on the path.");
+  params.addParam<bool>(
+      "variable_activation",
+      false,
+      "Whether to use variable value for element activation. If false, use path activation");
+  params.addParam<Real>("activate_value", 0.0, "The value above which to activate the element");
+  params.addCoupledVar(
+      "coupled_var",
+      "The variable value will be used to decide wether an element whould be activated.");
+
+  return params;
+}
+
+ActivateElementUserObject::ActivateElementUserObject(const InputParameters & parameters)
+  : ElementUserObject(parameters),
+    _active_subdomain_id(declareRestartableData<subdomain_id_type>(
+        "active_subdomain_id", getParam<subdomain_id_type>("active_subdomain_id"))),
+    _inactive_subdomain_id(declareRestartableData<subdomain_id_type>(
+        "inactive_subdomain_id", getParam<subdomain_id_type>("inactive_subdomain_id"))),
+    _expand_boundary_name(getParam<std::vector<BoundaryName>>("expand_boundary_name")),
+    _variable_activation(
+        declareRestartableData<bool>("variable_activation", getParam<bool>("variable_activation"))),
+    _function_x(isParamValid("function_x") ? &getFunction("function_x") : nullptr),
+    _function_y(isParamValid("function_y") ? &getFunction("function_y") : nullptr),
+    _function_z(isParamValid("function_z") ? &getFunction("function_z") : nullptr),
+    _activate_distance(
+        declareRestartableData<Real>("activate_distance", getParam<Real>("activate_distance"))),
+    _coupled_var(isParamValid("coupled_var") ? &coupledValue("coupled_var") : nullptr),
+    _activate_value(
+        declareRestartableData<Real>("activate_value", getParam<Real>("activate_value")))
+{
+  if (_variable_activation && _coupled_var == nullptr)
+    paramError("variable_activation",
+               "Need to specify a 'coupled_var' if variable_activation = true");
+
+  if ((!_variable_activation) && _coupled_var != nullptr)
+    mooseWarning("Not using variable activation, so the 'coupled_var' is ignored");
```

Comment:
all the awkward bool logic stuff becomes unnecessary if you split this into one class per functionality.